src/pydap/handlers/dap.py

```diff
@@ -134,6 +134,8 @@ def dataset_from_dap4(self):
                 self.fragment,
             )
         )
+        dmr_url = dmr_url.replace("[", "%5B").replace("]", "%5D")
```

Comment:
python's `request.urlparse` and `request.urlunparse` can split an url into 6 different component just like `urllib` in that Gemini example. That already was taking place before the new encoding feature of `[` and `]`. 

Here is the workflow of `open_url`:
1) (<ins>**Unchanged**</ins>) Pass the `url`  to the `DAPHandler` class, where it gets parsed and unparsed to determine the protocol (dap2 vs dap4). Relevant code here:

> https://github.com/pydap/pydap/blob/15afa3d0f2731965fcdfc89909bde3e46872452e/src/pydap/handlers/dap.py#L77-L98

The relevant `scheme`, `netloc`, `path`, `query`, etc of the `url` are assigned to `self.` so they become callable within the class. Note that the original `url` is now assigned to `base_url` and also becomes callable.

2) (<ins>**Changed**</ins>)`make_dataset` (line 96 above) creates a `DatasetType` with dummy variables. This `DatasetType` must have the right number of variables according to the query/constraint in the url. The important piece here is the `dmr_url`. Here is where `pydap` needs to escape the `[` `]` to generate the `dmr_url` accurately. For that, this PR takes the dmr_url and escapes '[' and ''].

For example, the test url generated from the DAP Access Form
```python
url = "http://test.opendap.org/opendap/hyrax/data/nc/test.nc?dap4.ce=/s33[0:][0:1:2];/s113[0:1:0][0:1:0][0:1:2]"
```
now produces the following escaped `drm_url`:

```python
"http://test.opendap.org/opendap/hyrax/data/nc/test.nc.dmr?dap4.ce=/s33%5B0:%5D%5B0:1:2%5D;/s113%5B0:1:0%5D%5B0:1:0%5D%5B0:1:2%5D"
```
If you  copy-paste into a browser, it takes you to the DAP Access Form which contains the right metadata.

3) (<ins>**Changed**</ins>) `add_proxies()`  (line 98 above) adds a proxy to the DatasetType  per variable excluding lat, lon, for the remote data set url.
<int>The change I did was to not append the variable `query`to the `ce` string</ins>. You can see that in the PR. The reason behind this was because of "constraint expreesion parse errors". 
Here is the original relevant code, which takes place within a loop per each variable in the DatasetType:
```python
# this is within a loop per each variable named self.id
scheme, netloc, path, params, query, fragment = urlparse(self.baseurl)
ce = "dap4.ce=" + quote(self.id) + hyperslab(index) + query
```
When there is no constraint, `query=''` and it adds nothing to ce. 
When there is a constraint,  `query` contains the constraint of the original file (`base_url`). This messes up the constraint expression for each `self.id` variable within the loop.

If query is NOT removed, I get the following error:

```python
HTTPError: 400 400
<?xml version="1.0" encoding="UTF-8"?>
<dap4:Error xmlns:dap4="http://xml.opendap.org/ns/DAP/4.0#" httpcode="400">
  <dap4:Message>libdap error transmitting DAP4Data: Constraint expression parse error: syntax error, unexpected word, expecting end of file</dap4:Message>
  <dap4:Context [/](http://localhost:8888/)>
  <dap4:OtherInformation [/](http://localhost:8888/)>
<[/dap4](http://localhost:8888/dap4):Error>
```

And so query string of the original `base_url` should not be added to `ce`.

Reply:
I will look at other potential characters that should be appropriately escaped, and include proper example of desired behavior that this PR should incorporate.